[SPARK-23319][TESTS][BRANCH-2.3] Explicitly specify Pandas and PyArrow versions in PySpark tests (to skip or test)

HyukjinKwon opened the pull request:
This PR backports https://github.com/apache/spark/pull/20487 to branch-2.3.

HyukjinKwon committed:
[SPARK-23319][TESTS] Explicitly specify Pandas and PyArrow versions in PySpark tests (to skip or test)

This PR proposes to explicitly specify Pandas and PyArrow versions in PySpark tests to skip or test.

We declared the extra dependencies:

https://github.com/apache/spark/blob/b8bfce51abf28c66ba1fc67b0f25fe1617c81025/python/setup.py#L204

In case of PyArrow:

Currently we only check if pyarrow is installed or not without checking the version. It already fails to run tests. For example, if PyArrow 0.7.0 is installed:

```
======================================================================
ERROR: test_vectorized_udf_wrong_return_type (pyspark.sql.tests.ScalarPandasUDF)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/.../spark/python/pyspark/sql/tests.py", line 4019, in test_vectorized_udf_wrong_return_type
    f = pandas_udf(lambda x: x * 1.0, MapType(LongType(), LongType()))
  File "/.../spark/python/pyspark/sql/functions.py", line 2309, in pandas_udf
    return _create_udf(f=f, returnType=return_type, evalType=eval_type)
  File "/.../spark/python/pyspark/sql/udf.py", line 47, in _create_udf
    require_minimum_pyarrow_version()
  File "/.../spark/python/pyspark/sql/utils.py", line 132, in require_minimum_pyarrow_version
    "however, your version was %s." % pyarrow.__version__)
ImportError: pyarrow >= 0.8.0 must be installed on calling Python process; however, your version was 0.7.0.

----------------------------------------------------------------------
Ran 33 tests in 8.098s

FAILED (errors=33)
```

In case of Pandas:

There are few tests for old Pandas which were tested only when Pandas version was lower, and I rewrote them to be tested when both Pandas version is lower and missing.

Manually tested by modifying the condition:

```
test_createDataFrame_column_name_encoding (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 1.19.2 must be installed; however, your version was 0.19.2.'
test_createDataFrame_does_not_modify_input (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 1.19.2 must be installed; however, your version was 0.19.2.'
test_createDataFrame_respect_session_timezone (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 1.19.2 must be installed; however, your version was 0.19.2.'
```

```
test_createDataFrame_column_name_encoding (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 0.19.2 must be installed; however, it was not found.'
test_createDataFrame_does_not_modify_input (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 0.19.2 must be installed; however, it was not found.'
test_createDataFrame_respect_session_timezone (pyspark.sql.tests.ArrowTests) ... skipped 'Pandas >= 0.19.2 must be installed; however, it was not found.'
```

```
test_createDataFrame_column_name_encoding (pyspark.sql.tests.ArrowTests) ... skipped 'PyArrow >= 1.8.0 must be installed; however, your version was 0.8.0.'
test_createDataFrame_does_not_modify_input (pyspark.sql.tests.ArrowTests) ... skipped 'PyArrow >= 1.8.0 must be installed; however, your version was 0.8.0.'
test_createDataFrame_respect_session_timezone (pyspark.sql.tests.ArrowTests) ... skipped 'PyArrow >= 1.8.0 must be installed; however, your version was 0.8.0.'
```

```
test_createDataFrame_column_name_encoding (pyspark.sql.tests.ArrowTests) ... skipped 'PyArrow >= 0.8.0 must be installed; however, it was not found.'
test_createDataFrame_does_not_modify_input (pyspark.sql.tests.ArrowTests) ... skipped 'PyArrow >= 0.8.0 must be installed; however, it was not found.'
test_createDataFrame_respect_session_timezone (pyspark.sql.tests.ArrowTests) ... skipped 'PyArrow >= 0.8.0 must be installed; however, it was not found.'
```

Author: hyukjinkwon <gurwls223@gmail.com>

Closes #20487 from HyukjinKwon/pyarrow-pandas-skip.

(cherry picked from commit 71cfba04aeec5ae9b85a507b13996e80f8750edc)

Signed-off-by: hyukjinkwon <gurwls223@gmail.com>

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #87165 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87165/testReport)** for PR 20534 at commit [`ff9ba5e`](https://github.com/apache/spark/commit/ff9ba5eb840bcd843c5201e23589e8cbb5009c53).

SparkQA commented:
**[Test build #87165 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87165/testReport)** for PR 20534 at commit [`ff9ba5e`](https://github.com/apache/spark/commit/ff9ba5eb840bcd843c5201e23589e8cbb5009c53).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87165/
Test PASSed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #87180 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87180/testReport)** for PR 20534 at commit [`ff9ba5e`](https://github.com/apache/spark/commit/ff9ba5eb840bcd843c5201e23589e8cbb5009c53).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/680/
Test PASSed.

ueshin commented:
LGTM.

ueshin commented:
@HyukjinKwon You can include the fix #20538 or backport it after this is merged. It's up to you.

HyukjinKwon commented:
Yea, will deal with it. Thanks for the reminder!

SparkQA commented:
**[Test build #87180 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87180/testReport)** for PR 20534 at commit [`ff9ba5e`](https://github.com/apache/spark/commit/ff9ba5eb840bcd843c5201e23589e8cbb5009c53).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87180/
Test FAILed.

HyukjinKwon commented:
Let me just pick up the followup here.

ueshin committed:
[SPARK-23319][TESTS][FOLLOWUP] Fix a test for Python 3 without pandas.

## What changes were proposed in this pull request?

This is a followup pr of #20487.

When importing module but it doesn't exists, the error message is slightly different between Python 2 and 3.

E.g., in Python 2:

```
No module named pandas
```

in Python 3:

```
No module named 'pandas'
```

So, one test to check an import error fails in Python 3 without pandas.

This pr fixes it.

## How was this patch tested?

Tested manually in my local environment.

Author: Takuya UESHIN <ueshin@databricks.com>

Closes #20538 from ueshin/issues/SPARK-23319/fup1.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/686/
Test PASSed.

SparkQA commented:
**[Test build #87185 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87185/testReport)** for PR 20534 at commit [`c110e34`](https://github.com/apache/spark/commit/c110e34f0137a031bcff602d6855b0b28febe9ab).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87185/
Test PASSed.
